axlearn/audio/asr_decoder.py

```diff
@@ -196,19 +196,30 @@ def _compute_target_paddings(
         target_labels: Tensor = input_batch["target_labels"]
         # Infer target_paddings from out-of-range labels.
         target_paddings = jnp.logical_or(cfg.vocab_size <= target_labels, target_labels < 0)
+        return target_paddings
 
+    def _input_stats_summary(
+        self, input_batch: Nested[Tensor]
+    ) -> Dict[str, Union[WeightedScalar, Tensor]]:
+        target_labels: Tensor = input_batch["target_labels"]
+        target_paddings = self._compute_target_paddings(target_labels)
         batch_size = target_labels.shape[0]
         target_lengths = jnp.sum(1 - target_paddings, axis=-1)
-        self.add_summary(
-            "input_stats/average_target_length",
-            WeightedScalar(jnp.mean(target_lengths), batch_size),
-        )
         source_lengths = jnp.sum(1 - input_batch["paddings"], axis=-1)
-        self.add_summary(
-            "input_stats/average_source_length",
-            WeightedScalar(jnp.mean(source_lengths), batch_size),
-        )
-        return target_paddings
+        # pytype: disable=attribute-error
+        ret_dict = {
+            "input_stats/average_target_length": WeightedScalar(
+                jnp.mean(target_lengths), batch_size
+            ),
+            "input_stats/average_source_length": WeightedScalar(
+                jnp.mean(source_lengths), batch_size
+            ),
+            "input_stats/frame_packing_effiency": WeightedScalar(
+                jnp.sum(source_lengths) / input_batch["paddings"].size, input_batch["paddings"].size
```

Comment:
Guard against division by 0 here and below?

Reply:
I think this before: the denominator `input_batch['paddings'].size` is directly pulled from the input. If it is zero, many other places will become problematic well before we hit here (e.g., various normalization layer). Nevertheless, I have added safeguard for them.

Reply:
Thanks! I don't think we should make assumptions about the other components.